src/utils/parse.rs

```diff
@@ -83,3 +84,286 @@ where
     }
     Ok(vector)
 }
+
+#[derive(PartialEq, Debug)]
+enum ParseState {
+    Start,
+    LeftBracket,
+    Index,
+    Value,
+    Splitter,
+    Comma,
+    Length,
+}
+
+#[inline(always)]
+pub fn svector_filter_nonzero<T: Zero + Clone + PartialEq>(
+    indexes: &[u32],
+    values: &[T],
+) -> (Vec<u32>, Vec<T>) {
+    let non_zero_indexes: Vec<u32> = indexes
+        .iter()
+        .enumerate()
+        .filter(|(i, _)| values.get(*i).unwrap() != &T::zero())
+        .map(|(_, x)| *x)
+        .collect();
+    let non_zero_values: Vec<T> = indexes
+        .iter()
+        .enumerate()
+        .filter(|(i, _)| values.get(*i).unwrap() != &T::zero())
+        .map(|(i, _)| values.get(i).unwrap().clone())
+        .collect();
+    (non_zero_indexes, non_zero_values)
+}
```

Comment:
You can just use two pointers to copy the data and call `truncate` to truncate the two arrays.